oximeter/db/src/client/mod.rs

```diff
@@ -4185,4 +4185,242 @@ mod tests {
         db.cleanup().await.unwrap();
         logctx.cleanup_successful();
     }
+
+    // The schema directory, used in tests. The actual updater uses the
+    // zone-image files copied in during construction.
+    const SCHEMA_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/schema");
+
+    #[tokio::test]
+    async fn check_actual_schema_upgrades_are_valid_single_node() {
+        check_actual_schema_upgrades_are_valid_impl(false).await;
+    }
+
+    #[tokio::test]
+    async fn check_actual_schema_upgrades_are_valid_replicated() {
+        check_actual_schema_upgrades_are_valid_impl(true).await;
+    }
+
+    // NOTE: This does not actually run the upgrades, only checks them for
+    // validity.
+    async fn check_actual_schema_upgrades_are_valid_impl(replicated: bool) {
+        let name = format!(
+            "check_actual_schema_upgrades_are_valid_{}",
+            if replicated { "replicated" } else { "single_node" }
+        );
+        let logctx = test_setup_log(&name);
+        let log = &logctx.log;
+
+        // We really started tracking the database version in 2. However, that
+        // set of files is not valid by construction, since we were just
+        // creating the full database from scratch as an "upgrade". So we'll
+        // start by applying version 3, and then do all later ones.
+        const FIRST_VERSION: u64 = 3;
+        for version in FIRST_VERSION..=OXIMETER_VERSION {
+            let upgrade_file_contents = Client::read_schema_upgrade_sql_files(
+                log, false, version, SCHEMA_DIR,
```

Comment:
It appears we're only testing single node and not replicated here?

Reply:
Derp (on me). Good find. I set `false` to `replicated` in my branch: https://github.com/oxidecomputer/omicron/pull/5897, which runs a new series of migrations (and why @bnaecker added this), and things are passing correctly. That PR is almost ready for merge, so, I'll ensure this test fn is updated.

Reply:
awesome, thanks!